crates/cli/src/subcommands/generate/csharp.rs

```diff
@@ -42,17 +42,31 @@ fn scalar_or_string_name(b: &AlgebraicType) -> Option<&str> {
 fn ty_fmt<'a>(ctx: &'a GenCtx, ty: &'a AlgebraicType, namespace: &'a str) -> impl fmt::Display + 'a {
     fmt_fn(move |f| match ty {
         AlgebraicType::Sum(sum_type) => {
-            // This better be an option type
-            if let Some(inner_ty) = sum_type.as_option() {
-                write!(f, "{}?", ty_fmt(ctx, inner_ty, namespace))
+            if sum_type.is_special() {
```

Comment:
Looking at those `if is_... { if let Some(...) ...` I get an impression that we would be better served by `as_special()` + an enum instead of a simple `is_special() -> bool`.

Then you could use it with a much cleaner code like

```
match ty.as_special() {
  Some(SpecialType::Identity) => ...,
  Some(SpecialType::Option(ty)) => ...,
  ...
  None => ...handle non-special types...
}
```

Reply:
At that point we could bite the bullet and add `TypespaceForGenerate`, which should express exactly the typespaces that satisfy `is_valid_for_client_code_generation`.